### PR TITLE
History Archive Support for State Archival

### LINKED ISSUE #4593: Add History Archive support for Hot Archive Bucket

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves #4593

This adds support for the Hot Archive BucketList in History Archives. This includes changes to publish, catchup, and assume state. 

~TODO: More robust p23 upgrade testing, specifically around upgrades that occur on or during checkpoint boundaries.~

Upgrade tests have been added, so unit tests should have complete coverage now.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

**Review by graydon — APPROVED** [review]
Looks good! Handful of nits but you can skip 'em / do in followups at your leisure.

File: src/bucket/BucketManager.cpp [hunk 1]
```diff
@@ -1094,15 +1112,17 @@ BucketManager::startBackgroundEvictionScan(uint32_t ledgerSeq,
         mSnapshotManager->copySearchableLiveBucketListSnapshot();
     auto const& sas = cfg.stateArchivalSettings();
 
-    using task_t = std::packaged_task<EvictionResultCandidates()>;
+    using task_t =
+        std::packaged_task<std::unique_ptr<EvictionResultCandidates>()>;
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/bucket/BucketManager.cpp:1116]: I threw a bunch of consts in `EvictionResultCandidates`, which removed the default copy constructor and copy assignment operator, which I need in `BucketManager.cpp::1154`. Eviction candidates can be pretty large, so we probably don't want to copy them anyway. I think the correct solution is to make `EvictionResultCandidates` and use `unique_ptr` everywhere.

**Comment by SirTyson** [comment]
This has been rebased and should be ready for review

File: src/history/test/HistoryTests.cpp [hunk 2]
```diff
@@ -174,57 +174,158 @@ TEST_CASE("History bucket verification", "[history][catchup]")
...
+}
+SECTION("verify fails hash mismatch"){SECTION("live buckets"){
+    liveHashes.push_back(bucketGenerator.generateBucket<LiveBucket>(
+        TestBucketState::HASH_MISMATCH));
+auto verify = wm.executeWork<DownloadBucketsWork>(
+    buckets, hotBuckets, liveHashes, hotHashes, *tmpDir);
+REQUIRE(verify->getState() == BasicWork::State::WORK_FAILURE);
+}
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
```
Comment (dmkozh) [inline src/history/test/HistoryTests.cpp:287]: Does this need to be compile-time-gated?

File: src/catchup/IndexBucketsWork.cpp [hunk 3]
```diff
@@ -86,8 +90,7 @@ IndexBucketsWork::IndexWork::postWork()
 
             if (!self->mIndex)
             {
-                // TODO: Fix this when archive BucketLists assume state
-                self->mIndex = createIndex<LiveBucket>(
+                self->mIndex = createIndex<BucketT>(
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/catchup/IndexBucketsWork.cpp:93]: Previously, this path could only index live Buckets and have the LiveBucket type hardcoded. It has now been "fixed" to be generic and work with both BucketList types when we assume state.

File: src/history/test/HistoryTestsUtils.cpp [hunk 4]
```diff
@@ -124,36 +144,50 @@ RealGenesisTmpDirHistoryConfigurator::configure(Config& mCfg,
...
     // Finish writing and close the bucket file
-    REQUIRE(mBuf);
-    mOut.writeOne(*mBuf, &mHasher, &mBytesPut);
-    mObjectsPut++;
-    mBuf.reset();
-    mOut.close();
-
-    return std::pair<std::string, uint256>(mFilename.string(),
-                                           mHasher.finish());
+    REQUIRE(this->mBuf);
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/history/test/HistoryTestsUtils.cpp:183]: It is required, at least on clang-18. I think its because the base class is templated, so you need to explicitly use this since members could be dependent on template type.

File: src/history/test/HistoryTests.cpp (hunk 2 above)
Comment (SirTyson) [reply src/history/test/HistoryTests.cpp:287]: Yes, reason being this test manually calls Hot Archive functions that are otherwise protocol gated. Without the ifdef, we call these in p22 curr, which fails asserts.

File: src/history/test/HistoryTestsUtils.cpp [hunk 5]
```diff
@@ -493,10 +534,13 @@ CatchupSimulation::generateRandomLedger(uint32_t version)
                                10;
             res.writeBytes = 100'000;
             uint32_t inclusion = 100;
+
             sorobanTxs.push_back(createUploadWasmTx(
-                getApp(), stroopy, inclusion, DEFAULT_TEST_RESOURCE_FEE, res));
+                getApp(), stroopy, inclusion, DEFAULT_TEST_RESOURCE_FEE * 5,
```
[the remark this answers is not in the document]
Comment (SirTyson) [reply src/history/test/HistoryTestsUtils.cpp:539]: The fee increased since we need to submit more transactions/close more ledgers in order to populate the Hot Archive, which increases dynamic fees such as rent and write fees. I'll update the constant.
